File: src/main/domains/agents/knowledge-base-store.ts
import { app } from 'electron'
import { existsSync, mkdirSync, readFileSync, readdirSync, rmSync, writeFileSync } from 'fs'
import { join } from 'path'
import type { AgentKnowledgeDoc, KnowledgeBaseConfig } from '@shared/agents'

type KnowledgeBaseSeed = Pick<KnowledgeBaseConfig, 'id' | 'name'> &
  Partial<Pick<KnowledgeBaseConfig, 'docCount' | 'indexedAt'>>

type DocumentSeed = Omit<AgentKnowledgeDoc, 'kbId'> & Partial<Pick<AgentKnowledgeDoc, 'kbId'>>

export interface KnowledgeBaseStore {
  listKnowledgeBases(): KnowledgeBaseConfig[]
  saveKnowledgeBase(input: KnowledgeBaseSeed): KnowledgeBaseConfig
  deleteKnowledgeBase(id: string): void
  listDocuments(kbId: string): AgentKnowledgeDoc[]
  saveDocument(kbId: string, input: DocumentSeed): AgentKnowledgeDoc
  deleteDocument(kbId: string, docId: string): void
}

function knowledgeBasesRootDir(baseDir?: string): string {
  return baseDir?.trim() || join(app.getPath('userData'), 'agents', 'knowledge-bases')
}

function ensureDir(dir: string): void {
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true })
  }
}

function indexFilePath(rootDir: string): string {
  return join(rootDir, 'index.json')
}

function kbDir(rootDir: string, kbId: string): string {
  return join(rootDir, kbId)
}

function docsDir(rootDir: string, kbId: string): string {
  return join(kbDir(rootDir, kbId), 'docs')
}

function docFilePath(rootDir: string, kbId: string, docId: string): string {
  return join(docsDir(rootDir, kbId), `${docId}.json`)
}

function normalizeKnowledgeBaseSummary(input: unknown): KnowledgeBaseConfig | null {
  if (!input || typeof input !== 'object') return null
  const raw = input as Record<string, unknown>
  const id = typeof raw['id'] === 'string' ? raw['id'].trim() : ''
  const name = typeof raw['name'] === 'string' ? raw['name'].trim() : ''
  if (!id || !name) return null
  return {
    id,
    name,
    docCount: typeof raw['docCount'] === 'number' ? raw['docCount'] : 0,
    indexedAt: typeof raw['indexedAt'] === 'number' ? raw['indexedAt'] : null
  }
}

function normalizeKnowledgeDoc(input: unknown, kbId: string): AgentKnowledgeDoc | null {
  if (!input || typeof input !== 'object') return null
  const raw = input as Record<string, unknown>
  const id = typeof raw['id'] === 'string' ? raw['id'].trim() : ''
  const title = typeof raw['title'] === 'string' ? raw['title'].trim() : ''
  const content = typeof raw['content'] === 'string' ? raw['content'] : ''
  if (!id || !title) return null
  return {
    id,
    kbId,
    title,
    content,
    createdAt: typeof raw['createdAt'] === 'number' ? raw['createdAt'] : Date.now(),
    updatedAt: typeof raw['updatedAt'] === 'number' ? raw['updatedAt'] : Date.now()
  }
}

function readKnowledgeBaseIndex(rootDir: string): KnowledgeBaseConfig[] {
  ensureDir(rootDir)
  const filePath = indexFilePath(rootDir)
  if (!existsSync(filePath)) return []
  try {
    const raw = readFileSync(filePath, 'utf-8')
    const parsed = JSON.parse(raw) as unknown
    if (!Array.isArray(parsed)) return []
    return parsed
      .map((item) => normalizeKnowledgeBaseSummary(item))
      .filter((item): item is KnowledgeBaseConfig => item !== null)
  } catch {
    return []
  }
}

function writeKnowledgeBaseIndex(rootDir: string, knowledgeBases: KnowledgeBaseConfig[]): void {
  ensureDir(rootDir)
  writeFileSync(indexFilePath(rootDir), JSON.stringify(knowledgeBases), 'utf-8')
}

function updateKnowledgeBaseSummary(
  rootDir: string,
  kbId: string,
  updater: (current: KnowledgeBaseConfig) => KnowledgeBaseConfig
): KnowledgeBaseConfig {
  const knowledgeBases = readKnowledgeBaseIndex(rootDir)
  const current = knowledgeBases.find((item) => item.id === kbId)
  if (!current) {
    throw new Error(`知识库不存在: ${kbId}`)
  }
  const next = updater(current)
  const updated = knowledgeBases.map((item) => (item.id === kbId ? next : item))
  writeKnowledgeBaseIndex(rootDir, updated)
  return next
}

export function createKnowledgeBaseStore(baseDir?: string): KnowledgeBaseStore {
  const rootDir = knowledgeBasesRootDir(baseDir)

  function listKnowledgeBases(): KnowledgeBaseConfig[] {
    return readKnowledgeBaseIndex(rootDir)
  }

  function saveKnowledgeBase(input: KnowledgeBaseSeed): KnowledgeBaseConfig {
    ensureDir(rootDir)
    ensureDir(kbDir(rootDir, input.id))
    ensureDir(docsDir(rootDir, input.id))

    const knowledgeBases = readKnowledgeBaseIndex(rootDir)
    const current = knowledgeBases.find((item) => item.id === input.id)
    const next: KnowledgeBaseConfig = {
      id: input.id,
      name: input.name.trim(),
      docCount: current?.docCount ?? input.docCount ?? 0,
      indexedAt: current?.indexedAt ?? input.indexedAt ?? null
    }
    const updated = current
      ? knowledgeBases.map((item) => (item.id === next.id ? next : item))
      : [...knowledgeBases, next]
    writeKnowledgeBaseIndex(rootDir, updated)
    return next
  }

  function deleteKnowledgeBase(id: string): void {
    const knowledgeBases = readKnowledgeBaseIndex(rootDir).filter((item) => item.id !== id)
    writeKnowledgeBaseIndex(rootDir, knowledgeBases)
    rmSync(kbDir(rootDir, id), { recursive: true, force: true })
  }

  function listDocuments(kbId: string): AgentKnowledgeDoc[] {
    const dir = docsDir(rootDir, kbId)
    ensureDir(dir)
    try {
      return readdirSync(dir)
        .filter((file) => file.endsWith('.json'))
        .map((file) => {
          const raw = readFileSync(join(dir, file), 'utf-8')
          return normalizeKnowledgeDoc(JSON.parse(raw) as unknown, kbId)
        })
        .filter((item): item is AgentKnowledgeDoc => item !== null)
        .sort((a, b) => a.updatedAt - b.updatedAt)
    } catch {
      return []
    }
  }

  function saveDocument(kbId: string, input: DocumentSeed): AgentKnowledgeDoc {
    ensureDir(kbDir(rootDir, kbId))
    ensureDir(docsDir(rootDir, kbId))
    const now = Date.now()
    const next: AgentKnowledgeDoc = {
      id: input.id.trim(),
      kbId,
      title: input.title.trim(),
      content: input.content,
      createdAt: typeof input.createdAt === 'number' ? input.createdAt : now,
      updatedAt: typeof input.updatedAt === 'number' ? input.updatedAt : now
    }
    writeFileSync(docFilePath(rootDir, kbId, next.id), JSON.stringify(next), 'utf-8')
    updateKnowledgeBaseSummary(rootDir, kbId, (current) => ({
      ...current,
      docCount: listDocuments(kbId).length
    }))
    return next
  }

  function deleteDocument(kbId: string, docId: string): void {
    rmSync(docFilePath(rootDir, kbId, docId), { force: true })
    updateKnowledgeBaseSummary(rootDir, kbId, (current) => ({
      ...current,
      docCount: listDocuments(kbId).length
    }))
  }

  return {
    listKnowledgeBases,
    saveKnowledgeBase,
    deleteKnowledgeBase,
    listDocuments,
    saveDocument,
    deleteDocument
  }
}

export function migrateLegacyKnowledgeBases(
  legacyKnowledgeBases: unknown[],
  store: KnowledgeBaseStore
): KnowledgeBaseConfig[] {
  for (const item of legacyKnowledgeBases) {
    if (!item || typeof item !== 'object') continue
    const raw = item as Record<string, unknown>
    const id = typeof raw['id'] === 'string' ? raw['id'].trim() : ''
    const name = typeof raw['name'] === 'string' ? raw['name'].trim() : ''
    if (!id || !name) continue
    store.saveKnowledgeBase({
      id,
      name,
      indexedAt: typeof raw['indexedAt'] === 'number' ? raw['indexedAt'] : null
    })
    const docs = Array.isArray(raw['docs']) ? raw['docs'] : []
    for (const doc of docs) {
      const normalized = normalizeKnowledgeDoc(doc, id)
      if (!normalized) continue
      store.saveDocument(id, normalized)
    }
  }
  return store.listKnowledgeBases()
}

File: server/app/domains/agents/knowledge_base_store.py
"""
知识库文档存储

对应 TypeScript 端 knowledge-base-store.ts
"""

import json
from pathlib import Path
from typing import Optional

from app.models.agents import AgentKnowledgeDoc, KnowledgeBaseConfig


class KnowledgeBaseStore:
    """知识库存储 —— 管理知识库元数据和文档"""

    def __init__(self, base_dir: str) -> None:
        self._root = Path(base_dir) / "agents" / "knowledge-bases"
        self._root.mkdir(parents=True, exist_ok=True)
        self._index_file = self._root / "index.json"

    def _read_index(self) -> list[dict]:
        """读取知识库索引文件"""
        if not self._index_file.exists():
            return []
        try:
            raw = self._index_file.read_text(encoding="utf-8")
            data = json.loads(raw)
            return data if isinstance(data, list) else []
        except (json.JSONDecodeError, ValueError):
            return []

    def _write_index(self, items: list[dict]) -> None:
        """写入知识库索引文件"""
        self._index_file.write_text(json.dumps(items, ensure_ascii=False), encoding="utf-8")

    def _docs_dir(self, kb_id: str) -> Path:
        return self._root / kb_id / "docs"

    def _doc_path(self, kb_id: str, doc_id: str) -> Path:
        return self._docs_dir(kb_id) / f"{doc_id}.json"

    # ---- 知识库 CRUD ----

    def list_knowledge_bases(self) -> list[KnowledgeBaseConfig]:
        """列出所有知识库"""
        items = self._read_index()
        return [KnowledgeBaseConfig.model_validate(item) for item in items]

    def save_knowledge_base(self, kb: KnowledgeBaseConfig) -> KnowledgeBaseConfig:
        """创建或更新知识库"""
        items = self._read_index()
        found = False
        for item in items:
            if item.get("id") == kb.id:
                item["name"] = kb.name
                kb = KnowledgeBaseConfig.model_validate(item)
                found = True
                break
        if not found:
            items.append({
                "id": kb.id,
                "name": kb.name,
                "docCount": kb.doc_count,
                "indexedAt": kb.indexed_at,
            })
        self._write_index(items)
        return kb

    def update_knowledge_base(self, kb_id: str, patch: dict) -> KnowledgeBaseConfig:
        """部分更新知识库配置"""
        items = self._read_index()
        for item in items:
            if item.get("id") == kb_id:
                item.update(patch)
                self._write_index(items)
                return KnowledgeBaseConfig.model_validate(item)
        raise ValueError(f"知识库 {kb_id} 不存在")

    def delete_knowledge_base(self, kb_id: str) -> None:
        """删除知识库及其所有文档"""
        items = [item for item in self._read_index() if item.get("id") != kb_id]
        self._write_index(items)
        kb_dir = self._root / kb_id
        if kb_dir.exists():
            import shutil
            shutil.rmtree(kb_dir, ignore_errors=True)

    # ---- 文档 CRUD ----

    def list_documents(self, kb_id: str) -> list[AgentKnowledgeDoc]:
        """列出知识库下的所有文档"""
        docs_dir = self._docs_dir(kb_id)
        if not docs_dir.exists():
            return []
        docs = []
        for f in sorted(docs_dir.glob("*.json")):
            try:
                raw = f.read_text(encoding="utf-8")
                data = json.loads(raw)
                data["kb_id"] = kb_id
                docs.append(AgentKnowledgeDoc.model_validate(data))
            except (json.JSONDecodeError, ValueError):
                continue
        return docs

    def save_document(self, kb_id: str, doc: AgentKnowledgeDoc) -> AgentKnowledgeDoc:
        """保存文档"""
        doc.kb_id = kb_id
        docs_dir = self._docs_dir(kb_id)
        docs_dir.mkdir(parents=True, exist_ok=True)
        self._doc_path(kb_id, doc.id).write_text(
            doc.model_dump_json(ensure_ascii=False), encoding="utf-8"
        )
        return doc

    def delete_document(self, kb_id: str, doc_id: str) -> bool:
        """删除文档"""
        try:
            self._doc_path(kb_id, doc_id).unlink()
            return True
        except FileNotFoundError:
            return False
